packages/flutter_tools/lib/src/commands/upgrade.dart

```diff
@@ -29,64 +38,157 @@ class UpgradeCommand extends FlutterCommand {
     printStatus('Upgrading Flutter from ${Cache.flutterRoot}...');
-
     await ChannelCommand.upgradeChannel();
+  }
 
-    int code = await runCommandAndStreamOutput(
-      <String>['git', 'pull', '--ff-only'],
+  /// Attempts to rebase the upstream onto the local branch.
+  ///
+  /// If there haven't been any hot fixes or local changes, this is equivalent
+  /// to a fast-forward.
+  Future<void> attemptRebase() async {
+    final int code = await runCommandAndStreamOutput(
+      <String>['git', 'pull', '--rebase'],
       workingDirectory: Cache.flutterRoot,
       mapFunction: (String line) => matchesGitLine(line) ? null : line,
     );
-
-    if (code != 0)
+    if (code != 0) {
       throwToolExit(null, exitCode: code);
+    }
+  }
 
-    // Check for and download any engine and pkg/ updates.
-    // We run the 'flutter' shell script re-entrantly here
-    // so that it will download the updated Dart and so forth
-    // if necessary.
+  /// Update the engine repository.
+  ///
+  /// Check for and download any engine and pkg/ updates.
+  /// We run the 'flutter' shell script re-entrantly here
+  /// so that it will download the updated Dart and so forth
+  /// if necessary.
+  Future<void> updateEngine() async {
     printStatus('');
     printStatus('Upgrading engine...');
-    code = await runCommandAndStreamOutput(
+    final int code = await runCommandAndStreamOutput(
       <String>[
         fs.path.join('bin', 'flutter'), '--no-color', '--no-version-check', 'precache',
       ],
       workingDirectory: Cache.flutterRoot,
       allowReentrantFlutter: true,
     );
+    if (code != 0) {
+      throwToolExit(null, exitCode: code);
+    }
+  }
 
+  /// Update the user's packages.
+  Future<void> updatePackages(FlutterVersion flutterVersion) async {
     printStatus('');
     printStatus(flutterVersion.toString());
-
     final String projectRoot = findProjectRoot();
     if (projectRoot != null) {
       printStatus('');
       await pubGet(context: PubContext.pubUpgrade, directory: projectRoot, upgrade: true, checkLastModified: false);
     }
+  }
 
-    // Run a doctor check in case system requirements have changed.
+  /// Run flutter doctor in case requirements have changed.
+  Future<void> runDoctor() async {
     printStatus('');
     printStatus('Running flutter doctor...');
-    code = await runCommandAndStreamOutput(
+    await runCommandAndStreamOutput(
       <String>[
         fs.path.join('bin', 'flutter'), '--no-version-check', 'doctor',
       ],
       workingDirectory: Cache.flutterRoot,
       allowReentrantFlutter: true,
     );
+  }
 
-    return null;
+  /// Pop stash changes.
+  ///
+  /// This should only be called if [maybeStashChanges] returned true.
+  Future<void> popChanges() async {
+    try {
+      await runCheckedAsync(<String>[
+        'git', 'stash', 'pop',
+      ]);
+    } catch (e) {
+      printError('Failed to re-apply local changes. State may have been lost.');
```

Comment:
Rather than popping, let's make sure they still have their stash that we created for them just in case.

What about having `maybeStashChanges()` return the stash name String (or `null` if no stash was saved)?  Then this would become `applyStash(String stashName)`.